src/RevCPU.cc

```diff
@@ -2437,12 +2521,246 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
+            AssignedThreads.at(ProcID).erase(AssignedThreads.at(ProcID).begin()+HartID);
+            CompletedThreads.emplace(Thread->GetThreadID());
+            break;
+          case ThreadState::BLOCKED:
+            // This thread is blocked (currently only caused by a rev_pthread_join)
+            // We need to: 
+            // 1. Check if the thread it is waiting on has already completed
+            // 2. If it has... Thread can resume execution
+            // 3. If not, thread remains blocked
+            //    3a. Move its ThreadID to the BlockedThreads list
+            //    3b. Remove it from the AssignedThreads list
+            output.verbose(CALL_INFO, 8, 0, "Thread %u on Core %u is BLOCKED\n", Thread->GetThreadID(), ProcID);
+            // -- 1.
+            if( ThreadCanProceed(Thread->GetThreadID()) ){
+              // -- 2.
+              output.verbose(CALL_INFO, 8, 0, "Thread %u on Core %u was waiting on thread %u which has already completed so it can proceed\n",
+                             Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
+              // Continue executing thread on same Core
+              Thread->SetState(ThreadState::RUNNING);
+            }
+            else { // -- 3.
+              output.verbose(CALL_INFO, 8, 0, "Thread %u on Core %u was waiting on thread %u which has not yet completed so it remains blocked\n",
+                             Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
+              Thread->SetState(ThreadState::BLOCKED);
+              // -- 3a.
+              BlockedThreads.emplace(Thread->GetThreadID());
+              // -- 3b.
+              AssignedThreads.at(ProcID).erase(AssignedThreads.at(ProcID).begin()+HartID);
+            }
+            break;
+          case ThreadState::START: // Should never happen
+            output.fatal(CALL_INFO, 99, "Error: Thread %u on Core %u is assigned but is in START state... This is a bug\n",
+                         Thread->GetThreadID(), ProcID);
+            break;
+          case ThreadState::RUNNING: 
+            output.verbose(CALL_INFO, 11, 0, "Thread %u on Core %u is RUNNING\n", Thread->GetThreadID(), ProcID);
+            break;
+          case ThreadState::READY: 
+            // If this happens we are not setting state when assigning thread somewhere 
+            output.fatal(CALL_INFO, 99, "Error: Thread %u on Core %u is assigned but is in START state... This is a bug\n",
+                         Thread->GetThreadID(), ProcID);
+            break;
+          default: // Should DEFINITELY never happen
+            output.fatal(CALL_INFO, 99, "Error: Thread %u on Core %u is in an unknown state... This is a bug\n", 
+                         Thread->GetThreadID(), ProcID);
+            break;
+        }
+        // State change has been handled, so clear the bit
+        Changes[HartID] = false;
+      }
```

Comment:
Some questions about this switch:
 - Can a Proc have more than one thread in RUNNING state at once? If it does, is that an error condition?
 - This statement transitions from BLOCKED to RUNNING - is this the only place that is possible?  The crux of the question is are there multiple places the threadstate can be updated, and is there potential for conflict, or multiple state changes in a single clock

Reply:
As of now, RevProc will only ever change a RevThread's state to `DONE` or `BLOCKED`. The rest of State management is handled in RevCPU.